ADR/0027-availability-probe-framework.md

```diff
@@ -0,0 +1,102 @@
+labels to allow differentiating between the different probes. When exported out of the
+cluster of origin, additional labels will be attached to the metric to mark the cluster
+in which the metric was generated.
+
+### Delivery Method
+
+The availability metric will be computed for each component based on the exit status of
+the latest execution of a CronJob evaluating the component's availability. The different
+teams will provide the CronJobs' implementation, while the RHTAP Observability team will
+provide the means to translate the jobs' results to the required metric.
+
+It is up for each team to define what it means for its component(s) to be available.
+Fundamentally, a component should be reported as available as long as it's capable of
+providing the service it aims to provide, and unavailable otherwise.
+
+Each team will define CronJobs that will test the availability of their components.
+The Job started by each CronJob will terminate successfully if the test complete
+successfully, and will terminate with an error in case the test fails.
+
+Kube-state-metrics generates Jobs and CronJob metrics that can be processed using
+Prometheus recording rules in order to generate RHTAP's availability metric.
+
+The RHTAP Observability team will construct the necessary rules.
+
+The resulting metric will identify each probe based on its CronJob name and the
+namespace in which the CronJob was defined:
+
+* The name of the CronJob will be translated to a label denoting the name of the probe.
+* The namespace in which the CronJob is defined will be translated to a label denoting
+  the probe's namespace.
+
+#### Probes Requirements and Design Considerations
+
+A Job running for a probe is required to:
+
+* Evaluate the availability of the component being probed.
+* Exit with status `Failed` if the component was evaluated to be unavailable or with
+  status `Complete` if it was evaluated to be available (referring to the status field
+  of the Kubernetes Job resource).
+* Be able to exhaust its `backoffLimit` by the time it's due to run again (e.g. if the
+  cronjob is defined to run every 10 minutes and can take up to 2 minutes to execute
+  it cannot have a `backoffLimit` larger than 4).
+* Clean up all resources generated during its run.
+
+Considerations for defining probes' CronJobs:
+
+* Where should it run in order to provide reliable evaluation?
+    * Which namespace?
+    * Which cluster?
+* What sort of permissions does it require to have?
+    * Would that be a good-enough representation of what it aims to evaluate?
+* What sort of APIs does it need to access?
+* How often should it run?
+* Does it affect performance in a reasonable manner?
+* How resources are to be cleaned up?
+    * Upon startup? completion? failure?
+    * Using another CronJob? Finalizers?
+
+#### Naming Convention
+
+To allow generating the metric only for relevant CronJobs, probe CronJob names should
+have a standardized format:
+
+* `rhtap-probe-<probe_name>`
```

Comment:
I would say that it should be better to use labels instead of names -> for caching/filtering options.

Reply:
I agree. We have some limitations that do not allow us going that way: The Prometheus instance collecting the job metrics does not necessarily collect labels used for cronjobs (for cardinality reasons, the specific labels need be whitelisted inside the Prometheus configs, which we're not the owners for).